internal/workerpool/pool.go

```diff
@@ -30,7 +30,7 @@ type Task func(context.Context)
 // longer ask for more tasks, and will return after waiting for the running
 // goroutines to finish. The provided context can be used to cancel everything
 // and exit the loop.
-func Run(ctx context.Context, limit int, nextTask func(context.Context) Task) {
+func Run(ctx context.Context, limit int, nextTask func(context.Context) Task) error {
```

Comment:
The error-handing/cancelation pattern that you do in ReceiveConcurrently is actually more subtle that what you're doing here. Would you consider putting that logic here? So both nextTask and Task would return error, and cause Run to exit with that error after waiting for outstanding goroutines.

Reply:
Seems reasonable. I'll try it.

Reply:
Done. PTAL.